36 comtradebaci importer

mharoruiz opened the pull request:
This PR adds a `BACI` importer class to `bblocks_data_importers` with the following changes:

- The importer lives in `bblocks_data_importers/cepii/baci.py`
- Test module for the importer is in `tests/test_cepii/test_baci.py`
- A doc called `baci.ipynb` documents the importer's basic functionality

mharoruiz committed:
Create __init__.py

mharoruiz committed:
Create baci.py

mharoruiz committed:
Update __init__.py

mharoruiz committed:
Update config.py

mharoruiz committed:
Add documentation

mharoruiz committed:
Update config.py

mharoruiz committed:
Update pyarrow

mharoruiz committed:
Optimize file storage

mharoruiz committed:
Improve parquet handling

mharoruiz committed:
Incorporate data validation

mharoruiz committed:
Add tests

mharoruiz committed:
black .

mharoruiz committed:
Update documentation

mharoruiz committed:
Add baci doc to index.md

mharoruiz committed:
Add test docstrings

mharoruiz committed:
black .

mharoruiz committed:
Update config.py

mharoruiz committed:
pass hs to get_data, add iso3 and name optionally

mharoruiz committed:
Separate static methods

mharoruiz committed:
Update tests

mharoruiz committed:
black .

mharoruiz commented:
In addition to the previously mentioned changes, the PR now includes:
- `hs_version` is passed to `get_data()` data.
- iso3 codes are included in the final df by default. Users can decided whether they include country names as well
- Statics methods are included in a separate module `file_handling.py`

lpicci96 reviewed (changes requested):
@mharoruiz Overall, this is a well thought out importer. There are a few key design issues we should discuss. For the future we should really start with a design document before jumping into development.

__Important points__:

1. __State Mutation and Object Design__

The current implementation involves internal state mutation—specifically, the `_data `and `_hs_version` attributes are overwritten each time `get_data()` is called. In our initial design discussions, we intended for the object to be flexible and not tied to a single HS version, allowing users to load and work with data across multiple versions and years within the same instance. However, the current design contradicts that vision by requiring an HS version at instantiation and mutating internal state based on subsequent get_data calls. This can lead to subtle bugs and makes reasoning about the object more difficult. I recommend choosing one of two clear directions:

1. Revert to a single-version-per-instance model, where the object is tied to a fixed HS version set at initialization.

2. Or go fully flexible and decouple the instance from any persistent HS version state, treating each get_data() call as a stateless query (with proper caching keyed by config).

For reference, the `weo` importer follows a design pattern that supports multiple configurations cleanly—worth reviewing for inspiration.

2.  __Caching Transparency__

While disk caching is appropriate given the size of the BACI dataset, its current implementation lacks transparency. The `clear_cache()` method only resets in-memory attributes (`_data`, `_metadata`) but leaves disk files intact, which could confuse users who expect a true reset. Moreover, because the object automatically overwrites state when `get_data()` is called, the value of `clear_cache()` is limited.

I suggest the following:

1. Require users to explicitly provide a data_path at instantiation, to reinforce that persistent storage is being used.

2. Add a `force_reload=True` argument to get_data() (and similar methods), allowing users to bypass cached files and fetch fresh data.

3. Expand `clear_cache()` to handle both memory and disk cleanup, e.g., def clear_cache(self, clear_disk: bool = False), and deprecate `delete_local_files()` in favour of this unified interface.



3 . __Fuzzy loading__

Our approach for the package is that a user should be able to call a getter method like `get_data` and there should be internal logic to figure out if the data has already been downloaded - if so serve the cached data, if not download it. Currently there are several getter methods like `get_metadata` which rely on having called `get_data`. We should avoid this and move that logic internally to the class.

__Other improvements__:

Documentation: Improve the module and class-level docstrings. The module docstring should include technical details helpful for maintainers (e.g., assumptions, known edge cases), while the class docstring should clearly explain usage patterns and expected behaviors. Also, document the structure and units of the data itself—e.g., is value in thousands of USD? Is quantity in metric tons? Assumptions about this are not obvious and should be made explicit.

Code structure: Several functions are doing too much. For instance, _get_latest_version() combines network I/O, HTML parsing, and string matching. Similarly, _download_and_extract() mixes downloading, decompression, and folder setup. These should be broken into smaller, more testable functions with single responsibilities.

Encapsulation: Some parsing and formatting utilities don't need to be part of the public class. Consider moving them to helper modules or making them static/private.

Constants: Move all hardcoded strings (e.g., URLs, filename patterns, HS version strings) to top-level, capitalized constants. For example: BASE_URL = "https://www.cepii.fr/DATA_DOWNLOAD/baci/data".

mharoruiz commented on src/bblocks_data_importers/cepii/baci.py:
```diff
@@ -0,0 +1,466 @@
+
+        if baci_version == "latest":
+            self._baci_version = self._get_latest_version()
+        else:
```
Actually `baci_version` can only be set to a literal string, as defined above:

```
baci_version: Literal[
            "202102", "202201", "202301", "202401", "202401b", "202501", "latest"
        ] = "latest",
```

Should I still define `ALLOWED_VERSIONS`?

lpicci96 replied:
The implementation of Literal won't prevent any other string from being passed, it is just for type checkers. See the [docs](https://docs.python.org/3/library/typing.html#typing.Literal). So there should be a way to tell the user if the version they requested is not valid. Now the approach of hard coding the allowed values might not be ideal because there will be new releases (like `202601`) so you'd have to update the code base every time there is a new version. It is better to programatically check if the version passed by the user exists. If it doesn't then raise an error.

mharoruiz committed:
Update BACI importer

mharoruiz committed:
Update static methods

mharoruiz committed:
Update documentation

mharoruiz committed:
Update baci.py

mharoruiz committed:
Update static_methods.py

mharoruiz committed:
Update tests

mharoruiz committed:
Improve errors

mharoruiz committed:
Update test_static_methods.py

mharoruiz committed:
black .

mharoruiz commented:
Thanks again for the thorough review! @lpicci96 

Given the rigidity of the BACI database, I updated the module following a single-version-per-instance model. I also tried to address all of your comments/suggestions.

Other than that, the most important changes are: 

### VERSIONS_DICT

Introduced `VERSIONS_DICT`, which is returned by `get_available_versions()`. This function scrapes the BACI website to return a dictionary that maps BACI versions to their supported HS versions. It also contains a bool "latest" variable to identify the latest BACI version. 

`VERSIONS_DICT` also serves to validate the BACI-HS version combination that the users passes. 

I thought it would be useful for the user to also have access to this dictionary, so I added a static method to the module called `get_versions()` which is just a wrapper method of `get_available_versions()`. 

### Separation of static methods

Added a `static_methods` module and populated it heavily with functions that support the module but are not strictly dependent on the importer class. 

Let me know what you think.

mharoruiz committed:
Merge remote-tracking branch 'origin/main' into 36-comtradebaci-importer

mharoruiz committed:
Update README.md

mharoruiz committed:
Change namespace

mharoruiz committed:
black

mharoruiz committed:
Resolve namespace issues

mharoruiz committed:
Improve coverage

mharoruiz committed:
Update documentation

mharoruiz committed:
Move get_baci_versions out of class

mharoruiz committed:
Update test_baci.py

mharoruiz committed:
Update baci.py

lpicci96 commented on docs/docs/importers/baci.md:
```diff
@@ -0,0 +1,88 @@
+## Data caching
+
+The data and metadata are cached to avoid loading the dataset repeatedly. Use the `clear_cache()` method to delete this
+data. You can set clear_disk = True to delete the local directory where the BACI data was saver (defaults to False).
```
wrap `clear_disk = True` in a code block

lpicci96 reviewed (changes requested):
@mharoruiz Looks good. One major change and a few minor changes

The `force_reload` only reloads from disk instead of also redownloading the data which it should do. The doc string for this argument should also be better phrased - "force a new download of the data instead of using any cached data"

Please also update the changelog with a minor version

mharoruiz committed:
Delete __init__.py

mharoruiz committed:
Update baci_utils.py

mharoruiz committed:
Update config.py

mharoruiz committed:
Update baci.md

mharoruiz committed:
Update baci.py

mharoruiz committed:
Update tests

mharoruiz commented on src/bblocks/data_importers/cepii/baci.py:
```diff
@@ -0,0 +1,416 @@
+        self,
+        include_country_names: bool = True,
+        years: int | list[int] | range | set[int] | None = None,
+        force_reload: bool = False,
```
Good catch!

mharoruiz committed:
Merge branch 'main' into 36-comtradebaci-importer

mharoruiz committed:
Fix file search bug, replace get_data with _load_data

mharoruiz committed:
Update tests

mharoruiz committed:
black .

lpicci96 committed:
refactor version logic

lpicci96 committed:
refactor versions module

lpicci96 committed:
refactor data management

lpicci96 committed:
refactor baci object

lpicci96 committed:
fix rel import bug

lpicci96 committed:
add year filter logic

lpicci96 committed:
add pyarrow ds for data caching

lpicci96 committed:
update docs

lpicci96 committed:
update logs

lpicci96 committed:
add get product descriptions

lpicci96 committed:
add available years method

lpicci96 committed:
add country and prod labels functionality

lpicci96 committed:
add product filtering

lpicci96 committed:
add get available countries

lpicci96 committed:
update docs

lpicci96 committed:
black and module docs

lpicci96 committed:
Merge pull request #66 from ONEcampaign/baci-refactor

Baci refactor

lpicci96 commented:
@mharoruiz back to you for testing and documentation

mharoruiz committed:
Update parse_readme

mharoruiz committed:
Merge pull request #72 from ONEcampaign/parse_readme-refactor

`parse_readme` refactor
